pkg/webhooks/pod_mutating_webhook.go

```diff
@@ -0,0 +1,140 @@
+	_, usingNodeSelectorStrategy := pod.Annotations[jobset.NodeSelectorStrategyKey]
+	if !usingExclusivePlacement || usingNodeSelectorStrategy {
+		return nil
+	}
+	return p.patchPod(ctx, pod)
+}
+
+// patchPod will only add exclusive affinites to pod index 0. For all pods besides pod index 0,
+// it will add a scheduling gate to prevent them from scheduling initially.
+func (p *podWebhook) patchPod(ctx context.Context, pod *corev1.Pod) error {
+	log := ctrl.LoggerFrom(ctx)
+	// If this pod is part of a JobSet which is usig the node selector strategy for exclusive
+	if pod.Annotations[batchv1.JobCompletionIndexAnnotation] == "0" {
+		log.Info("pod webhook: setting exclusive affinities for node selector exclusive placement strategy")
+		setExclusiveAffinities(pod)
+		return nil
+	} else {
+		log.Info("pod webhook: adding node affinity for follower pod for node selector exclusive placement strategy")
+		return p.setNodeSelector(ctx, pod)
+	}
+}
+
+func setExclusiveAffinities(pod *corev1.Pod) {
+	if pod.Spec.Affinity == nil {
+		pod.Spec.Affinity = &corev1.Affinity{}
+	}
+	if pod.Spec.Affinity.PodAffinity == nil {
+		pod.Spec.Affinity.PodAffinity = &corev1.PodAffinity{}
+	}
+	if pod.Spec.Affinity.PodAntiAffinity == nil {
+		pod.Spec.Affinity.PodAntiAffinity = &corev1.PodAntiAffinity{}
+	}
+	// Pod affinity ensures the pods of this job land on the same topology domain.
+	pod.Spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution = append(pod.Spec.Affinity.PodAffinity.RequiredDuringSchedulingIgnoredDuringExecution,
+		corev1.PodAffinityTerm{
+			LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
+				{
+					Key:      jobset.JobKey,
+					Operator: metav1.LabelSelectorOpIn,
+					Values:   []string{pod.Labels[jobset.JobKey]},
+				},
+			}},
+			TopologyKey:       pod.Annotations[jobset.ExclusiveKey],
+			NamespaceSelector: &metav1.LabelSelector{},
+		})
+	// Pod anti-affinity ensures exclusively this job lands on the topology, preventing multiple jobs per topology domain.
+	pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution = append(pod.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution,
+		corev1.PodAffinityTerm{
+			LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
+				{
+					Key:      jobset.JobKey,
+					Operator: metav1.LabelSelectorOpExists,
+				},
+				{
+					Key:      jobset.JobKey,
+					Operator: metav1.LabelSelectorOpNotIn,
+					Values:   []string{pod.Labels[jobset.JobKey]},
+				},
+			}},
+			TopologyKey:       pod.Annotations[jobset.ExclusiveKey],
+			NamespaceSelector: &metav1.LabelSelector{},
+		})
+}
+
+func (p *podWebhook) setNodeSelector(ctx context.Context, pod *corev1.Pod) error {
+	log := ctrl.LoggerFrom(ctx)
+	leaderPod, err := p.leaderPodForFollower(ctx, pod)
+	if err != nil {
+		log.Error(err, "finding leader pod for follower")
```

Comment:
May be in this case we set the exclusive pod anti-affinity terms?

Reply:
Actually here we just need to return without an error and without changing anything. Mutating webhook is called before the Validating one. So if the leader is not scheduled yet, we return and the validate will reject creating this follower.

Reply:
done